### PR TITLE
[TwigBridge] ensure compatibility with Twig 3.15

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | 
| License       | MIT

take the changes from twigphp/twig#4398 into account

### CODE REVIEW

File: src/Symfony/Bridge/Twig/TokenParser/DumpTokenParser.php [hunk 1]
```diff
@@ -40,7 +40,7 @@ public function parse(Token $token): Node
         }
         $this->parser->getStream()->expect(Token::BLOCK_END_TYPE);
 
-        return new DumpNode($this->parser->getVarName(), $values, $token->getLine(), $this->getTag());
+        return new DumpNode($this->getVarName(), $values, $token->getLine(), $this->getTag());
```
Comment (xabbuh) [inline src/Symfony/Bridge/Twig/TokenParser/DumpTokenParser.php]: @fabpot Do I see it correctly that this is the only deprecation-free solution on 5.4 where we cannot change the `DumpNode` class (we could maybe make it accept a `LocalVariable` object in 7.2)?

Comment (stof) [reply src/Symfony/Bridge/Twig/TokenParser/DumpTokenParser.php]: we can change it to accept either a string or a LocalVariable.

File: src/Symfony/Bridge/Twig/TokenParser/DumpTokenParser.php [hunk 2]
```diff
@@ -50,4 +50,9 @@ public function getTag(): string
     {
         return 'dump';
     }
+
+    private function getVarName(): string
+    {
+        return sprintf('__internal_%s', hash('sha256', uniqid(mt_rand(), true)));
```
Comment (stof) [inline src/Symfony/Bridge/Twig/TokenParser/DumpTokenParser.php]: generating a random var name breaks reproducible template compilation, while we worked hard to make them reproducible a few years ago.

File: src/Symfony/Bridge/Twig/Tests/Node/DumpNodeTest.php [hunk 3]
```diff
@@ -74,11 +75,11 @@ public function testOneVar()
     {
         if (class_exists(Nodes::class)) {
             $vars = new Nodes([
-                new NameExpression('foo', 7),
+                class_exists(ContextVariable::class) ? new ContextVariable('foo', 7) : new NameExpression('foo', 7),
             ]);
         } else {
             $vars = new Node([
-                new NameExpression('foo', 7),
+                class_exists(ContextVariable::class) ? new ContextVariable('foo', 7) : new NameExpression('foo', 7),
```
Comment (stof) [inline src/Symfony/Bridge/Twig/Tests/Node/DumpNodeTest.php]: this is dead code. ContextVariable cannot exist if Nodes does not exist

File: src/Symfony/Bridge/Twig/NodeVisitor/TranslationDefaultDomainNodeVisitor.php [hunk 4]
```diff
@@ -51,8 +53,8 @@ public function enterNode(Node $node, Environment $env): Node
                 return $node;
             } else {
                 $var = $this->getVarName();
```
Comment (stof) [inline src/Symfony/Bridge/Twig/NodeVisitor/TranslationDefaultDomainNodeVisitor.php:55]: I just discovered that this node visitor breaks reproducible compilation here.

File: src/Symfony/Bridge/Twig/TokenParser/DumpTokenParser.php (hunk 1 above)
Comment (xabbuh) [reply src/Symfony/Bridge/Twig/TokenParser/DumpTokenParser.php]: I considered that too but I was wondering if doing that in a patch release is really the way to move forward.

**Comment by xabbuh** [comment]
I updated the `DumpNode` class to also accept a `LocalVariable` instance.

Though I am stuck at how we can deal with a similar case in the `StopwatchNode`. Any idea how to handle it?

File: src/Symfony/Bridge/Twig/NodeVisitor/TranslationDefaultDomainNodeVisitor.php (hunk 4 above)
Comment (xabbuh) [reply src/Symfony/Bridge/Twig/NodeVisitor/TranslationDefaultDomainNodeVisitor.php:55]: good point, let's fix this explicitly in a separate PR

File: src/Symfony/Bridge/Twig/TokenParser/DumpTokenParser.php (hunk 1 above)
Comment (xabbuh) [reply src/Symfony/Bridge/Twig/TokenParser/DumpTokenParser.php]: I did the change

**Comment by xabbuh** [comment]
I think I got a solution for the `StopwatchNode` as well.

Status: Needs Review

File: src/Symfony/Bridge/Twig/NodeVisitor/TranslationDefaultDomainNodeVisitor.php (hunk 4 above)
Comment (xabbuh) [reply src/Symfony/Bridge/Twig/NodeVisitor/TranslationDefaultDomainNodeVisitor.php:55]: I just noticed that this was fixed for 7.2 in #57609. Should we backport that PR?

Comment (stof) [reply src/Symfony/Bridge/Twig/NodeVisitor/TranslationDefaultDomainNodeVisitor.php:55]: that PR is actually broken

Comment (nicolas-grekas) [reply src/Symfony/Bridge/Twig/NodeVisitor/TranslationDefaultDomainNodeVisitor.php:55]: let's backport (and fix the broken part) in 5.4 yes
reproducible builds FTW

**Comment by nicolas-grekas** [comment]
Thank you @xabbuh.

File: src/Symfony/Bridge/Twig/NodeVisitor/TranslationDefaultDomainNodeVisitor.php (hunk 4 above)
Comment (stof) [reply src/Symfony/Bridge/Twig/NodeVisitor/TranslationDefaultDomainNodeVisitor.php:55]: A way to generate a reproducible name that does not clash with parent or child template could be to use a hash of the template name in the variable name (with the new hardcoded name as prefix)

Comment (xabbuh) [reply src/Symfony/Bridge/Twig/NodeVisitor/TranslationDefaultDomainNodeVisitor.php:55]: see #58706

**Pull request merged**